### ISSUE TITLE
Explanation for low ETH hashrate on maxwell 2.0 cards?

### ISSUE THREAD

**Issue by RhynarAI** [Lolliedieb/lolMiner-releases#604]
Hi there,

I am trying out several older cards. 
When using Tesla M40 (Maxwell 2.0, 12GB VRAM, 3xxx CUDA Cores) or Titan X (pretty mich the same card), the ETC Hashrate with lolminer 1.29 (under Hive OS or windows) drops down to around 2 MH/s. I have read several reports on the internet that this is common, the discussion often ending with "they are just old"....

But since I used these Cards with other compute tasks, I know they should be faster. So I did some comparisons with GTX 1650, GCN 1, 2, 4 Cards, etc.... And benched them against each other. In Vulkan, Open CL and Cuda Benchmarks, the Maxwell 2.0 Cards hold up quite nice, but drop down in Daggerhashimoto/ETH far below even RX550 4GB running in Zombie mode on a PCIE3/8x Connection (around 7MH/S).

To illustrate, just one example of geekbench results:
OPENCL Geekbench 5
TESLA M40 12GB: 32123 Points -> 2MH/s Daggerhashimoto ETH
RX550 4GB: 12302 Points -> 7MH/s Daggerhasmimoto ETH
GTX Titan X 12GB: 38855 Points -> 2.1 MH/s Daggerhashimoto ETH
GTX 1060: 33123 Points -> 20 MH/s ETH

So the MAxwell 2.0 12GB Cards are around the same speed as the Kepler 1060 in geekbench OPENCL, In Vulkan, the Maxwell even are much faster. In Cuda Geekbench, they are almost the same... Did a ton of different benchmarks (luxmar, tensorflow, computebench, gaming, etc...)- in everyone the 12GB MAxwell Cards are around the same as 1060 or faster...
tried out several motherboards, nvidia driver versions, cuda toolkit versions...

But in DAggerhashimito/ETH, the speed drops to around 1/10 of the 1060..

What is going on? Does anyone have a solid explanation or even a fix ?

**Comment by pdieppa** [comment]
I wonder if it has something to do with the memory bandwidth of the cards. Example, a 1080 ti has almost double the memory bandwidth of the Tesla M40.

This is a cool website that shows many specs for a bunch of older cards. I think last time it was updated was 2 years ago, so new models are not listed. 

https://www.gpuzoo.com/GPU-NVIDIA/index.html

**Comment by RhynarAI** [comment]
These specs can also be obtained from techpowerup and others, yes. SOme differences in architecture could explain differences, but we are talking about 1/10 of a speed for cards with similar speeds.
Of course I looked up the mem differencex you mention, too... A 1060 has the same GDDR5 Mem, a smaller Bus, a slightly higher clock count on the VRAM and is slower in almost every regards than the GTX Titan X / Tesla M40 - but still 10 times faster in DaggerHashimoto. There must be someting going on which is less obvious than simply "next generation speedups"

**Comment by pdieppa** [comment]
Have you tried other miners to see if the M40 still gets 2MH/s doing eth?

**Comment by RhynarAI** [comment]
I did try some out, but didn`t pay much attention back then to the M40 results (a lot of different testing with all kinds of older cards was done), so I might do that. But since I was reading many comments on the net from people recognizing the same behaviour, I suspected there is a general issue somehow here...

I just crawled through some release notes of lolminer and found that it explicitely states "maxwell and above support".. And a few versions ago, Fermi and Kepler Kernels were introduced. These are even older than Maxwell. So my thinking is: "If maxwell is simply too old (the reason why most other threads covering this topic stop), how come an extra Fermi and Kepler Kernel makes sense?.

I get the feeling, this should be solvable... I just cross-checked more benchmarks and spec comparisons and I really can´t find a good reason why the maxwell cards behave this slow..

Or is there some architectural oddity in maxwell (and only there) which conflicts with the way daggerhashimoto works?

**Comment by RhynarAI** [comment]
Tried some others (minerstat-OS Benchmark) on the 12GB Tesla M40 (maxwell 2.0) on Ethereum

Bminer: 2.5 MH/s
Claymore: Obviously fails (AFAIK Claymore is not usable anymore at all, no matter the card, right?)
ethminer: 2.5 MH/s (reports EPOCH 375..? not sure why)
GMiner: 2.315 MH/s (DAG creation speed was reported as 435 MB/S, taking roughly over 10sec.. normal?)
nanominer: crashed
lolminer: 2.27 MH/s
nfsminer: 2.49 MH/s, reporting Epoch 375 and generating a 3.99GB DAG
Phoenix: 2.3 MH/s, Epoch 417 (I think, screen scrolled too fast) with around 380 MB/s creation speed
T-REX: 2.31 MH/s, Epoch 417, DAG Creation 10sec 

I think it might be a good idea to check VRAM-Copy Speeds device-device. I´ve read about a bug on the GTX 970 which slowed down considerably above 3.5G of 4GB VRAM (reports of the 980 not being affected..). Maybe The Titan X and Tesla M40 have something similar? 
Could this be of help in chasing the cause:
https://github.com/kruzer/poclmembench
?
Seems to be used back then to identify the 970 speed bug..

**Comment by bunkaq** [comment]
I have M4000, P2000 and T4. 
M4000  1.8MH/s ETH
P2000 is doing 15.7 MH/s ETH
T4  27.5 MH/s  ETH (if you manage to cool it below 75C)

Maxwell cards are slow for ETH because of low memory clock and even core is ultra low. I am mining on M4000 rather Zano or VTC using it for ETH is not profitable at all. I was trying ERGO which is going about 20MH/s.

**Comment by RhynarAI** [comment]
Thanx for commenting.
Please take the time to read the above again :)
The "typical explanations" simply make no sense. If we take a look at the mentioned P2000 vs a Maxwell 2.0 Titan X, the Titan outperforms the P2000 in almost every regard by a large margin. In every other Game, Compute, etc.. benchmark, the Titan wins... 

So the explanation, why the Titan is ten times slower in ETH Mining than the Kepler card must lie somewhere else than simply saying "clocrates are slower" (which in many cases isn´t even the case, and even if they were, we are talkting a few percent...)

If you look up your P2000 and compare it to the maxwell Titan: VRAM Clock rates and type of memory are the same, the Titan even having a bigger bus, so mem-rates actually are better on the older card :)
The GPU is slightly lower clocked in max. boost, but the maxwell has a lot more cores, compute performance of maxwell titan is above your mentioned P2000. But still - the Pascal Card is 10 times faster.

Again: what still is missing is a good explanation, why a high end card from a previous generation, that is faster in every reagard than a midrange card from the follow up generation is around ten times slower in ETH mining... It can´t be as simple as "clock rate is slower" :)

**Comment by RhynarAI** [comment]
Another observation, that might give a clue is efficiency.... 
By chance, I stumpled across a GTX750TI (Maxwell1) and a GTX960 (Maxwell2). When using ZHASH on both, the 

750TI is 
around 0.33Mh/J with 8.2 sol/sek

960
around 0.23 MH/J with 13.6 Sol/Sek

Both have the same DRam Type, the 960 around double the cores... In this cae something weird is going on with efficiency - and to compare that to the TiTan/Tesla 12GB Maxwell 2.0 - it drops in efficiency way below the GTX 960.... (The 750 TI Actually coming close to some of smaller Pascal cards - in terms of efficiency...)

In the meantime, I ran a lot of other calculations on the Tesla/Titan and my inital thoughts still hold up: This card easily competes with mid range Pascal  Cards in many regards, I haven´t found one example where the Pascal card outperforms the Maxwell by a ratio of 1:10.... So I still think the difference in ETH Mining has to do something with the algo itself...
Since Kepler and Fermi Kernels now exist :
hm... just an idea - would building a "maxwell Kernel" be an idea? Maybe "cuda 5.2" built switches would adress maxwell specific issues?
